crates/goose/src/providers/formats/openai.rs

```diff
@@ -426,19 +443,63 @@ where
+            // Accept both SSE style "data: [DONE]" and plain "[DONE]" sentinel
+            if response.as_ref().is_ok_and(|s| s.trim() == "data: [DONE]" || s.trim() == "[DONE]") {
                 break 'outer;
             }
             let response_str = response?;
+            tracing::debug!(raw_stream_line = %response_str, "LLM_STREAM_RAW_LINE");
             let line = strip_data_prefix(&response_str);
 
             if line.is_none() || line.is_some_and(|l| l.is_empty()) {
                 continue
             }
 
-            let chunk: StreamingChunk = serde_json::from_str(line
-                .ok_or_else(|| anyhow!("unexpected stream format"))?)
-                .map_err(|e| anyhow!("Failed to parse streaming chunk: {}: {:?}", e, &line))?;
+            let line_str = line
+                .ok_or_else(|| anyhow!("unexpected stream format"))?;
+
+            // First parse into a generic JSON value so we can detect provider-specific
+            // error objects (some OpenAI-compatible endpoints emit {"object":"error",...}).
+            let v: Value = serde_json::from_str(line_str)
+                .map_err(|e| anyhow!("Failed to parse streaming JSON value: {}: {:?}", e, &line_str))?;
+
+            // Detect top-level error objects and yield a structured error message so the UI
+            // can render provider errors consistently during streaming. Previously we returned
+            // Err(...) which terminated the stream without providing a message object to the
+            // frontend; that made debugging custom providers difficult because the UI had no
+            // provider response to display.
+            if v.get("object").and_then(|o| o.as_str()).map(|s| s == "error").unwrap_or(false)
+                || v.get("error").is_some()
+            {
+                // Build a helpful text representation including status-like fields when available
+                let mut details = String::new();
+                // Attempt to include an explicit "message" field if present
+                if let Some(m) = v.get("message") {
+                    details.push_str(&format!("message: {}\n", m));
+                }
+                // Include any nested error object
+                if let Some(err_obj) = v.get("error") {
+                    details.push_str(&format!("error: {}\n", err_obj));
+                }
+                // Include the full JSON payload as a last resort
+                details.push_str(&format!("raw: {}", v));
+
+                // Construct a Message that contains the error details so the UI can display it.
+                // Use assistant role so it is shown where other assistant output appears, but
+                // include the error text so renderers can style it as an error.
+                let msg_text = format!("LLM streaming error encountered. See details below:\n{}", details);
+
+                let message = Message::assistant().with_content(MessageContent::text(msg_text));
```

Comment:
If we do need to show errors while keeping the stream alive (and I could see how that makes sense) we should have instead a non-terminating error type where we can return these without putting them in a Message